server/src/main/java/io/crate/expression/predicate/IsNullPredicate.java

```diff
@@ -151,7 +151,15 @@ public static Query refExistsQuery(Reference ref, Context context) {
                 return new FieldExistsQuery(field);
             }
             if (ArrayType.unnest(ref.valueType()) instanceof ObjectType objType) {
+                // This logic is applied for non-array object type columns as well since ArrayType.unnest(nonArrayType) == nonArrayType.
                 BooleanQuery.Builder booleanQuery = new BooleanQuery.Builder();
+                if (objType.innerTypes().isEmpty()) {
+                    // This is not required but just short circuit to fall into GenericFunctionQuery right away
```

Comment:
My understanding is that if we have an object (can be one some leaf object in a big object tree since we call this recursively) without inner types it means all values are either null or empty({} for objects or [] for arrays). Looking up in index `may` work but may not so instead of doing
```
Queries.not(refExistsQuery), Occur.MUST) //index
LuceneQueryBuilder.genericFunctionFilter(function, context), Occur.MUST) // full scan
```
we just do full scan.

Another question - since `refExistsQuery` is also used in `NotPredicate` and `ArrayUpperFunction` I would keep this isEmpty check and just remove 
 ```
if (ref.valueType() instanceof ArrayType<?>) {
                // For empty arrays there is nothing in the index and fieldsExists checks don't work
                return new BooleanQuery.Builder()
                    .add(Queries.not(refExistsQuery), Occur.MUST)
                    .add(LuceneQueryBuilder.genericFunctionFilter(function, context), Occur.MUST)
                    .build();
            }
```

why lookup I index if we say full scan is `MUST`?